sparse_strips/vello_common/src/strip.rs

```diff
@@ -204,23 +221,53 @@ pub fn render(
             (p1_y, p1_x, p0_y, p0_x)
         };
 
+        let (line_left_x, line_left_y, line_right_x) = if p0_x < p1_x {
+            (p0_x, p0_y, p1_x)
+        } else {
+            (p1_x, p1_y, p0_x)
+        };
+
         let y_slope = (line_bottom_y - line_top_y) / (line_bottom_x - line_top_x);
         let x_slope = 1. / y_slope;
 
-        {
-            // The y-coordinate of the intersections between the line and the tile's left and right
-            // edges respectively.
-            //
-            // There's some subtlety going on here, see the note on `line_px_left_y` below.
-            let line_tile_left_y = (line_top_y - line_top_x * y_slope)
-                .max(line_top_y)
-                .min(line_bottom_y);
-            let line_tile_right_y = (line_top_y + (Tile::WIDTH as f32 - line_top_x) * y_slope)
-                .max(line_top_y)
-                .min(line_bottom_y);
-
-            winding_delta +=
-                sign as i32 * ((line_tile_left_y <= 0.) != (line_tile_right_y <= 0.)) as i32;
+        winding_delta += sign as i32 * tile.winding as i32;
+
+        // TODO: this should be removed when out-of-viewport tiles are culled at the
+        // tile-generation stage. That requires calculating and forwarding winding to strip
+        // generation.
+        if tile.x == 0 && line_left_x < 0. {
+            let (ymin, ymax) = if line.p0.x == line.p1.x {
+                (line_top_y, line_bottom_y)
+            } else {
+                let line_viewport_left_y = (line_top_y - line_top_x * y_slope)
+                    .max(line_top_y)
+                    .min(line_bottom_y);
+
+                (
+                    f32::min(line_left_y, line_viewport_left_y),
+                    f32::max(line_left_y, line_viewport_left_y),
+                )
+            };
+
+            for y_idx in 0..Tile::HEIGHT {
+                let px_top_y = y_idx as f32;
+                let px_bottom_y = 1. + y_idx as f32;
+
+                let ymin = f32::max(ymin, px_top_y);
+                let ymax = f32::min(ymax, px_bottom_y);
+
+                let h = (ymax - ymin).max(0.);
+                accumulated_winding[y_idx as usize] += sign * h;
+
+                for x_idx in 0..Tile::HEIGHT {
```

Comment:
@tomcur should this have said Tile::WIDTH?

Reply:
Yes, thank you! Opened #848.